File: wwwroot/js/lobby.js
import startGame from "./main.js";
import requestAlert from "./alert.js";
import { colorPopup } from "./popupMenu.js";

let player;

// For color selection popup
let colors;
let associatedPlayers;

const colorSelectButton = document.getElementById("colorSelectButton");
const startGameButton = document.getElementById("startGameButton");
const lobbyIdText = document.getElementById("lobbyIdText");
const playerList = document.getElementById("playerList");

export default function joinLobby(id, p) {
    player = p;

    startGameButton.disabled = true;
    lobbyIdText.textContent = id;

    // Hide menu
    const menuContainer = document.getElementById('menu-container');
    menuContainer.style.display = 'none';

    // Un-hide lobby
    const lobbyContainer = document.getElementById('lobby-container');
    lobbyContainer.style.display = 'flex';

    var connection = new signalR.HubConnectionBuilder().withUrl("/lobbyHub?lobby=" + id).withAutomaticReconnect().build();
    setupConnection(connection, id);
}

function setupConnection(connection, id) {
    connection.start().then(() => {
        startGameButton.disabled = false;
    }).catch((err) => {
        return console.error(err.toString());
    });

    connection.on("MatchStart", () => {
        startGame(id, player);
    })

    connection.on("UpdatePlayerList", (players) => {
        playerList.innerHTML = "";
        players.forEach((p) => {
            var temp = document.createElement("li");
            temp.innerText = p.name;
            if (p.publicId === player.publicId) {
                temp.style.textDecoration = "underline";
                temp.style.fontWeight = "bold";
            }
            playerList.appendChild(temp);
        });
    })

    connection.on("FailedNotHost", () => {
        requestAlert("Only the host can start the match.");
    })

    connection.on("FailedNotEnoughPlayers", () => {
        requestAlert("There aren't enough players to start a match.");
    })

    startGameButton.addEventListener("click", (e) => {
        connection.invoke("StartMatch", player.privateId).catch((err) => {
            return console.error(err.toString());
        });
    })

    connection.on("PossibleColors", (Colors) => {
        colors = Colors;
    })

    connection.on("FailedColorTaken", () => {
        requestAlert("That color is already taken.");
    })

    connection.on("UpdateColorState", (associatedPlayersUpdated) => {
        associatedPlayers = associatedPlayersUpdated;

        if (document.getElementById("color-popup-menu") !== null) {
            colorPopup(connection, player, colors, associatedPlayers);
        }
    })

    colorSelectButton.addEventListener("click", (e) => {
        colorSelectButton.blur();
        colorPopup(connection, player, colors, associatedPlayers);
    })
}

File: Assets/Scripts/Multiplayer/Lobby.cs
public class Lobby
{
    public string Id { get; }
    public List<Player> Players { get; } = new List<Player>();
    public Player? host = null;

    public Lobby() => Id = Utils.RandomId();

    public bool AddPlayer(string privateId, string publicId, string name)
    {
        if (Full) return false;
        if (Players.Any((player) => player.PrivateId == privateId || player.PublicId == publicId)) return false;

        int color = -1;
        for (int i = 0; i < Constants.Colors.Count(); i++)
        {
            if (!ColorTaken(i))
            {
                color = i;
                break;
            }
        }

        if (color == -1) return false;

        Player player = new Player(privateId, publicId, name, color);

        if (host is null) host = player;
        Players.Add(player);
        return true;
    }

    public bool Full => Players.Count >= Constants.MaximumPlayerCount;
    public bool ColorTaken(int color) => Players.Any(player => player.Color == color);
}